### PR TITLE
New Registration API

### PR DESCRIPTION
**Pull request by APickledWalrus**
### Description
This is built from the work done in #5331 (thanks @kiip1)

This PR is the initial effort to build a modern API for Skript. All new API has been marked as experimental. Backwards compatibility has been kept. The focus of this PR is to introduce a new addon registration process and syntax registration process. It also implements many new utility interfaces and classes for usage across Skript.

# Breaking Changes

- The SyntaxElementInfo getters in the Skript class have been updated to return unmodifiable lists.

# New Skript Core

The new [Skript](https://github.com/SkriptLang/Skript/blob/feature/registration-rework/src/main/java/org/skriptlang/skript/Skript.java) interface represents the core of the language. At this time, implementations require methods for registering an addon and obtaining a collection of all addons. The Skript interface is an extension of the SkriptAddon interface, which is further described below.

A default implementation is available, which is what the plugin uses. Below is an example of creating a Skript instance using the default implementation. 
```java
Skript skript = Skript.of(getClass(), "Skript"); // parameter is the name for the addon representing Skript
```
*NOTE: The class parameter represents the main class of the application registering the addon.*

# New Addon API

The [SkriptAddon](https://github.com/SkriptLang/Skript/blob/feature/registration-rework/src/main/java/org/skriptlang/skript/addon/SkriptAddon.java) interface is the center of the new Addon API. At this time, every SkriptAddon has a **name**, **syntax registry**, and **localizer** (the last two are described in further detail below)

An addon can be registered as follows:
```java
Skript skript = ...
SkriptAddon myAddon = skript.registerAddon(getClass(), "myAddon");
```

## Addon Modules

Also included in this new API are [AddonModules](https://github.com/SkriptLang/Skript/blob/feature/registration-rework/src/main/java/org/skriptlang/skript/addon/AddonModule.java), which enables Skript and its addons to be broken up into categories. For example, a Skript implementation not dependent on Minecraft might have a default module containing something like arithmetic.

Modules have two loading phases: `init` and `load`. `init` is intended for registering components that might be used by other modules (e.g. class infos). `load` is intended for registering components that are more specific to the module, such as syntax.
When loading multiple modules together, their `init` phases will first all be executed, and then their `load` phases will all be executed.

Loading a module is rather simple:
```java
SkriptAddon myAddon = ...
myAddon.loadModules(new MathModule())
```

# New Syntax Registration API

The syntax registration API is the biggest component of this PR. The registration process has been completely overhauled to provide greater control over the registration process.

## Syntax Infos

The [SyntaxInfo](https://github.com/SkriptLang/Skript/blob/feature/registration-rework/src/main/java/org/skriptlang/skript/registration/SyntaxInfo.java) interface is a replacement for the [SyntaxElementInfo](https://github.com/SkriptLang/Skript/blob/feature/registration-rework/src/main/java/ch/njol/skript/lang/SyntaxElementInfo.java) class.

Every SyntaxInfo has the following properties:
- An origin describing where it is from
- SyntaxElement class providing the implementation
- A method for obtaining a new instance of the SyntaxElement class
- A collection of uncompiled patterns (string form)
- A priority that dictates its position in the registry

Two [default extensions](https://github.com/SkriptLang/Skript/blob/feature/registration-rework/src/main/java/org/skriptlang/skript/registration/DefaultSyntaxInfos.java) exist:
- SyntaxInfo.Expression
  - Has an additional property for the return type of the expression
- SyntaxInfo.Structure
  - Has an additional property for the EntryValidator of the structure

Additionally, one Bukkit-specific implementation exists for Bukkit events:
- [BukkitInfos.Event](https://github.com/SkriptLang/Skript/blob/feature/registration-rework/src/main/java/org/skriptlang/skript/bukkit/registration/BukkitInfos.java) which has the following properties:
  - Documentation methods to be used for something like SimpleEvent
  - A collection of event classes for the Bukkit events the SkriptEvent represents

### Builders

Four builders exist for creating a SyntaxInfo, SyntaxInfo.Expression, SyntaxInfo.Structure, and BukkitInfos.Event. These builders are implemented using the new Builder/Buildable interfaces, described below.

The following is an example of building a SyntaxInfo for [LitPi](https://github.com/SkriptLang/Skript/blob/feature/registration-rework/src/main/java/ch/njol/skript/expressions/LitPi.java) with the builder for SyntaxInfo.Expression:
```java
SyntaxInfo.Expression.builder(LitPi.class, Double.class) // Syntax class, Return Type(s) superclass
	.origin(SyntaxOrigin.of(Skript.instance())) // (optional) This syntax is from Skript itself
	.supplier(LitPi::new) // (optional) Provides Skript with a way to create the object rather than using reflection
	.priority(SyntaxInfo.SIMPLE) // (optional) Equivalent to ExpressionType.SIMPLE
	.addPattern("(pi|π)")
	.build();
```

### SyntaxOrigin

A SyntaxOrigin describes where a syntax has come from. By default, it only has a name (String) which describes the origin. When specified, this would likely be the name of an addon (see [AddonOrigin](https://github.com/SkriptLang/Skript/blob/feature/registration-rework/src/main/java/org/skriptlang/skript/registration/SyntaxOrigin.java#L47)). When not specified, it is the fully qualified name of the SyntaxElement class.

### Priorities

The priority system is an evolution of [ExpressionType](https://github.com/SkriptLang/Skript/blob/feature/registration-rework/src/main/java/ch/njol/skript/lang/ExpressionType.java). Unlike ExpressionType, which is only for expressions, the priority system applies to all SyntaxInfos. By default, Skript has three priorities:
- SIMPLE (for patterns that are simple text)
- COMBINED (for patterns containing at least one expression) 
- PATTERN_MATCHES_EVERYTHING (for patterns, like ExprArithmetic, that are likely to (partially) match most user inputs)

`ExpressionType#EVENT` and `ExpressionType#PROPERTY` have been replaced with the constants `EventValueExpression#DEFAULT_PRIORITY` and `PropertyExpression#DEFAULT_PRIORITY` respectively. The former lies between SIMPLE/COMBINED while the latter lies between COMBINED/PATTERN_MATCHES_EVERYTHING
Additionally, `PropertyCondition#DEFAULT_PRIORITY` has been added for conditions. It also lies between COMBINED/PATTERN_MATCHES_EVERYTHING.

Unlike Structure priorities, this new Priority system is purely relational. There are no "magic" numbers that determine position. The definition for the default three may be useful for visualizing this:
```java
Priority SIMPLE = Priority.base(); // A base priority is one that has no relation to another.
Priority COMBINED = Priority.after(SIMPLE); // That is, it comes after SIMPLE
Priority PATTERN_MATCHES_EVERYTHING = Priority.after(COMBINED); // That is, it comes after COMBINED, which comes after SIMPLE
```

Additionally, a Priority may be created that comes *before* some other Priority. If one wanted to create a Priority for syntax that should be parsed **really early**, they might do something like:
```java
Priority REALLY_EARLY = Priority.before(SIMPLE);
```

## SyntaxRegistry

The [SyntaxRegistry](https://github.com/SkriptLang/Skript/blob/feature/registration-rework/src/main/java/org/skriptlang/skript/registration/SyntaxRegistry.java) is the home for an addon's syntax infos. A SyntaxRegistry has three important methods:
- `syntaxes(Key)` which returns all SyntaxInfos registered under a Key (e.g. all expressions, effects, etc.)
- `register(Key, SyntaxInfo)` which registers a SyntaxInfo under a Key
- `unregister(Key, SyntaxInfo)` which unregisters a SyntaxInfo that is stored under a Key

A default implementation of a SyntaxRegistry may be obtained through:
```java
SyntaxRegistry myRegistry = SyntaxRegistry.empty();
```

### Keys

Keys are used for storing SyntaxInfos of a certain type. Skript has six built in keys:
- `STRUCTURE` for structures
- `SECTION` for sections
- `STATEMENT` for effects and conditions (Statement classes)
- `EFFECT` for effects
- `CONDITION` for conditions
- `EXPRESSION` for expressions

Creating a Key constant is simple:
```java
Key<SyntaxInfo<? extends Statement>> STATEMENT = Key.of("statement");
```
Now, if we wanted to access the Statements in a registry:
```java
SyntaxRegistry registry = ...
registry.syntaxes(STATEMENT);
```

#### Child Keys
Child Keys are the same as Keys, but they also have a parent Key. Thus, when a SyntaxInfo is registered under a Child Key, it is also registered under the parent Key. This is how the register for Statements works, as the Effect and Condition Keys are Child Keys of the Statement key. We can use this example to see how to build Child Keys:
```java
Key<SyntaxInfo<? extends Statement>> STATEMENT = Key.of("statement");
// Anything registered under EFFECT will also be registered under STATEMENT
Key<SyntaxInfo<? extends Effect>> EFFECT = ChildKey.of(STATEMENT, "effect");
```

# Experimental Localization API

This PR includes an experimental (and likely subject to change) API for Localization. At this point in time, it exists for modern addons to register their language files. I avoided creating too much API as that will be for a separate PR reworking localization. The key idea here is loading language files, which can be done as follows:
```java
SkriptAddon addon = ...
addon.localizer().setSourceDirectories("lang", null);
```
`setSourceDirectories` takes in two parameters:
- `languageFileDirectory` which is the path to the language file directory on the jar.
- `dataFileDirectory` (optional) which is the path to the language file directory on the disk (e.g. user customizable lang files).

# New Utilities

## Builder/Buildable Interfaces
I have introduced two new interfaces for building objects along with converting objects into builders. The primary interface, builder, has two methods:
- `build()` which is the terminal operation for a builder that returns an instance of the type being built.
- `applyTo(Builder)` which enables applying (copying) the values of one builder onto another.

Another interface, Buildable, enables converting an object back into a builder. It has one method:
- `builder()` which returns a builder representing the object.

For example, one might want to add a new pattern to a SyntaxInfo. This is now trivial:
```java
SyntaxInfo info = ...;
info.builder()
	.addPattern("my new pattern")
	.build();
```

## ClassLoader API
I have introduced a ClassLoader utility class which acts as a replacement for the `SkriptAddon#loadClasses` method. It takes inspiration from the changes I had made in https://github.com/SkriptLang/Skript/pull/4573.

A simple utility method functioning the same as `SkriptAddon#loadClasses` exists too:
```java
public class MyAddon extends JavaPlugin {

  public void onEnable() {
    // getFile() returns a File object representing the plugin's jar file
    ClassLoader.loadClasses(MyAddon.class, getFile(), "my.addon.skript", "expressions", "effects");
  }

}
```

However, a builder exists for creating custom ClassLoaders with different behavior:
```java
public class MyAddon extends JavaPlugin {

  public void onEnable() {
    // getFile() returns a File object representing the plugin's jar file
    ClassLoader.builder()
        .basePackage("my.addon.skript")
        .addSubPackages("expressions", "effects")
        .initialize(true) // sets whether classes should be initialized
        .deep(true) // sets whether subpackages of packages should be searched
        .forEachClass(clazz -> /*do something*/) // sets a consumer that is run on each initialized class.
        .build()
        .loadClasses(MyAddon.class, getFile());
  }

}
```
It is possible to load classes without passing `getFile()` (or a jar), though it is not preferred due to reliability reasons. It worked fine during my testing, but the documentation for the utility used (ClassPath from Guava) notes some potential issues.

## ViewProvider Interface
An interface has been added to represent objects that can have unmodifiable views of themselves created. An unmodifiable view allows read access into an object but prevents making any changes to its values. For example, an unmodifiable SkriptAddon would allow you to obtain properties such as its name, but it would prevent changes such as storing a new registry or loading a new module.

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:**
- #121 (this will be the beginning of a new Addon API)
- #6175 
- Likely others TBD as I find them. Feel free to add some.

### CODE REVIEW

File: src/main/java/org/skriptlang/skript/registration/SyntaxRegistry.java
Comment (UnderscoreTud) [inline src/main/java/org/skriptlang/skript/registration/SyntaxRegistry.java:86]: Annotate with `@UnmodifiableView`?

**Review by UnderscoreTud — CHANGES_REQUESTED** [review]
This is a very major step to decoupling Skript from Bukkit, great work!

File: src/main/java/ch/njol/skript/SkriptAddon.java
Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/SkriptAddon.java:41]: IIRC that annotation is used for classes that are not intended to be extended by other applications but cannot be marked as final

Comment (UnderscoreTud) [reply src/main/java/ch/njol/skript/SkriptAddon.java:41]: Ah, I didn't realize you were overriding some of its methods. You can dismiss this comment

File: src/main/java/org/skriptlang/skript/SkriptImpl.java
Comment (Moderocky) [inline src/main/java/org/skriptlang/skript/SkriptImpl.java:94]: I think it might be nice to throw a little more information here (e.g. what's the class path of the existing addon? is it another copy of the same addon or a different one with the same name?)

Comment (Moderocky) [reply src/main/java/org/skriptlang/skript/SkriptImpl.java:94]: This having been said, is there a reason we actually need to prevent two addons from having the same name? Aside from debug feedback messages is there anything else that would be a problem?

File: src/main/java/org/skriptlang/skript/registration/DefaultSyntaxInfosImpl.java
Comment (Moderocky) [inline src/main/java/org/skriptlang/skript/registration/DefaultSyntaxInfosImpl.java:96]: You're making a mess with these type parameters, either leave some internal documentation or name them something nicer (we don't pay by the letter here)

File: src/main/java/org/skriptlang/skript/lang/Priority.java
Comment (Moderocky) [inline src/main/java/org/skriptlang/skript/lang/Priority.java:39]: This isn't a constant so you don't need to name it in upper case.

**Review by Moderocky — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/Skript.java
Comment (kiip1) [reply src/main/java/ch/njol/skript/Skript.java:1444]: The exceptions would be better than silently modifying the list, without any effect. Might be a good change to make.

File: src/main/java/org/skriptlang/skript/SkriptImpl.java
Comment (kiip1) [reply src/main/java/org/skriptlang/skript/SkriptImpl.java:94]: No reasons to prevent them from having the same name except from it being logical. It's good to make addons have unique names as it will make it possible to find addons by name instead of having to obtain a class reference.

File: src/main/java/org/skriptlang/skript/addon/SkriptAddon.java
Comment (kiip1) [inline src/main/java/org/skriptlang/skript/addon/SkriptAddon.java:59]: Do we want to link Skript to the file system by design? Wouldn't it be better if translations were handled by a separate class which can be accessed through the SkriptAddon instance? This new translations class would have capabilities for parsing from a folder or just manually adding them. This way you as the addon developer have full control over what's done, and aren't required to place files somewhere.

There should be some kind of basic "BukkitSkriptAddon" which handles a lot of common things, but that shouldn't mean the top SkriptAddon interface should be aware of that.

File: src/main/java/org/skriptlang/skript/Skript.java
Comment (kiip1) [inline src/main/java/org/skriptlang/skript/Skript.java:77]: The state system might be something to reconsider. Does Skript really need a global state? The only reason why it's here as far as I can see is because you don't want to add new syntaxes after scripts have been loaded, but at a certain point we might just make the script parse method take in a syntax registry snapshot, meaning that it's parsed with the current state of the registry. The Bukkit-Skript implementation might want to keep something like this though, as Bukkit does work with on enabled etc.

File: src/main/java/org/skriptlang/skript/addon/SkriptAddon.java
Comment (APickledWalrus) [reply src/main/java/org/skriptlang/skript/addon/SkriptAddon.java:59]: Maybe we'd want to introduce a LocalizedSkriptAddon interface that contains these methods? I do agree that these methods shouldn't be here on the top SkriptAddon interface. In the future, after a rework of the Language system parsing from sources other than files could be supported and the interface (in terms of its methods) could be re-examined. Separating it out for now would be good though

Comment (kiip1) [reply src/main/java/org/skriptlang/skript/addon/SkriptAddon.java:59]: Yeah that sounds good.

File: src/main/java/ch/njol/skript/lang/SyntaxElementInfo.java
Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/lang/SyntaxElementInfo.java:117]: From what I can tell, it is for dealing with the generics required for creating an ExpressionInfo

File: src/main/java/org/skriptlang/skript/registration/DefaultSyntaxInfosImpl.java
Comment (APickledWalrus) [reply src/main/java/org/skriptlang/skript/registration/DefaultSyntaxInfosImpl.java:102]: I have updated the documentation on the interfaces and inherited the Javadoc on the internal implementations - let me know if you would prefer to see more though.

**Review by sovdeeth — COMMENTED** [review]
It seems solid and a great step forward for Skript. I'd like to see an example of an addon using this new system before I approve, though, as I want to know if what I'm picturing in my head is actually how you'd use these classes.

File: src/main/java/ch/njol/skript/SkriptAddon.java
Comment (Moderocky) [inline src/main/java/ch/njol/skript/SkriptAddon.java:44]: You don't need to bother with this annotation if all of the constructors are package-private, nothing could extend it anyway.

File: src/main/java/org/skriptlang/skript/bukkit/registration/BukkitInfos.java
Comment (Moderocky) [inline src/main/java/org/skriptlang/skript/bukkit/registration/BukkitInfos.java:35]: I'm not a huge fan of this class name, do you think we could call it something a bit nicer?

File: src/main/java/org/skriptlang/skript/registration/SyntaxInfoImpl.java
Comment (sovdeeth) [inline src/main/java/org/skriptlang/skript/registration/SyntaxInfoImpl.java:46]: Do we want to allow syntaxes to be registered with no patterns (empty list)?

File: src/main/java/org/skriptlang/skript/registration/SyntaxRegisterImpl.java
Comment (sovdeeth) [inline src/main/java/org/skriptlang/skript/registration/SyntaxRegisterImpl.java:56]: What happens if a is before c, and b is after a? When a and b are compared, a has a syntax priority, but not related to b, so it hits the first `return 1` and gets put after, when it really should be before b. Am I missing some symmetry requirement?

**Review by sovdeeth — CHANGES_REQUESTED** [review]
Extremely good work, Pickle, just two more questions

File: src/main/java/org/skriptlang/skript/registration/SyntaxRegisterImpl.java
Comment (APickledWalrus) [reply src/main/java/org/skriptlang/skript/registration/SyntaxRegisterImpl.java:56]: yeah this needs reworked 😬

File: src/main/java/org/skriptlang/skript/bukkit/registration/BukkitInfos.java
Comment (APickledWalrus) [reply src/main/java/org/skriptlang/skript/bukkit/registration/BukkitInfos.java:34]: BukkitSyntaxInfos? I'm not too sure what else might work☹️

**Review by sovdeeth — APPROVED** [review]

**Review by Moderocky — APPROVED** [review]

**Review by UnderscoreTud — DISMISSED** [review]

File: src/main/java/org/skriptlang/skript/bukkit/registration/package-info.java
Comment (APickledWalrus) [reply src/main/java/org/skriptlang/skript/bukkit/registration/package-info.java:1]: these are old 😔

File: src/main/java/ch/njol/skript/conditions/base/PropertyCondition.java
Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/conditions/base/PropertyCondition.java]: Right now, the methods return the SyntaxInfo that was registered. It would be possible to construct a builder from that SyntaxInfo, change the priority, and then register it again. Of course, having to unregister and then register again is not ideal, but needing to change the priority is not exactly common.

An alternative approach could be to revamp these methods to just return a SyntaxInfo (and then you have to register it yourself), but I'm not sure if that's ideal (and it might conflict with the existing method). What do you think?

**Comment by JakeGBLP** [comment]
Should there be the possibility of modifying already existing elements? Some expressions like the "name" expression are very generic and being able to add to them as an addon developer would be awesome, some use cases would be skbee merging its merge component expression with skript's join string expression, adding xyz coordinates of non standard objects to the coordinate expression; this could open up a world of fixes for conflicts and it would be very cool to see it implemented, and if it is i will surely use it.

**Comment by APickledWalrus** [comment]
> Should there be the possibility of modifying already existing elements? Some expressions like the "name" expression are very generic and being able to add to them as an addon developer would be awesome, some use cases would be skbee merging its merge component expression with skript's join string expression, adding xyz coordinates of non standard objects to the coordinate expression; this could open up a world of fixes for conflicts and it would be very cool to see it implemented, and if it is i will surely use it.
> 
> This would in a sense act like "patches".

The ability to unregister, modify, and then re-register infos exists. What you want is covered in https://github.com/SkriptLang/Skript/pull/6728

**Review by Efnilite — APPROVED** [review]
merge this !!!

**Pull request merged**
